src/main/java/com/aws/greengrass/ggq/Exec.java

```diff
@@ -0,0 +1,466 @@
+                if (!process.waitFor(5, TimeUnit.SECONDS)) {
+                    (stderr == null ? stdout : stderr).accept("\n[TIMEOUT after InterruptedException]\n");
+                    process.destroyForcibly();
+                }
+                throw ie;
+            }
+            stderrc.join(5000);
+            stdoutc.join(5000);
+            return Optional.of(process.exitValue());
+        }
+        return Optional.empty();
+    }
+
+    /**
+     * Get the stdout and stderr output as a string.
+     *
+     * @return String of output.
+     * @throws InterruptedException if thread is interrupted while executing
+     * @throws IOException          if execution of the process fails to start
+     */
+    public String execAndGetStringOutput() throws InterruptedException, IOException {
+        StringBuilder sb = new StringBuilder();
+        Consumer<CharSequence> f = sb::append;
+        withOut(f).withErr(f).exec();
+        return sb.toString().trim();
+    }
+
+    public void background(IntConsumer cb) throws InterruptedException, IOException {
+        whenDone = cb;
+        exec();
+    }
+
+    @SuppressWarnings("PMD.NullAssignment")
+    void setClosed() {
+        if (!isClosed.get()) {
+            final IntConsumer wd = whenDone;
+            final int exit = process == null ? -1 : process.exitValue();
+            isClosed.set(true);
+            if (wd != null) {
+                wd.accept(exit);
+            }
+        }
+    }
+
+    public boolean isRunning() {
+        return process == null ? !isClosed.get() : process.isAlive();
+    }
+
+    /**
+     * Get associated process instance representing underlying OS process.
+     *
+     * @return process object.
+     */
+    public Process getProcess() {
+        return process;
+    }
+
+    @Override
+    public synchronized void close() throws IOException {
+        Process p = process;
+        if (p != null && p.isAlive()) p.destroyForcibly();
+    }
+
+    @Override
+    public String toString() {
+        return Utils.deepToString(cmds, 90).toString();
+    }
+
+    /**
+     * Sends the lines of an InputStream to a consumer in the background.
+     */
+    private class Copier extends Thread {
+        private final Consumer<CharSequence> out;
+        private final InputStream in;
+        private int nlines = 0;
+
+        Copier(InputStream i, Consumer<CharSequence> s) {
+            super("Copier");
+            in = i;
+            out = s;
+            // Set as a daemon thread so that it dies when the main thread exits
+            setDaemon(true);
+            if (whenDone != null) {
+                if (numberOfCopiers == null) {
+                    numberOfCopiers = new AtomicInteger(1);
+                } else {
+                    numberOfCopiers.incrementAndGet();
+                }
+            }
+        }
+        
+        public int getNlines() { return nlines; }
+
+        @Override
+        public void run() {
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

The cyclomatic complexity of this method is 12. By comparison, 98% of the methods in the CodeGuru reference dataset have a lower cyclomatic complexity. This indicates the method has a high number of decisions and it can make the logic difficult to understand and test. We recommend that you simplify this method or break it into multiple methods.